Data/ByteString.hs

```diff
@@ -442,17 +442,18 @@ append = mappend
 -- | /O(n)/ 'map' @f xs@ is the ByteString obtained by applying @f@ to each
 -- element of @xs@.
 map :: (Word8 -> Word8) -> ByteString -> ByteString
-map f (BS fp len) = unsafeDupablePerformIO $ unsafeWithForeignPtr fp $ \a ->
-    create len $ map_ 0 a
+map f (BS fp len) = unsafeDupablePerformIO $ unsafeWithForeignPtr fp $ \ptr1 ->
+    create len $ \ptr2 -> m ptr1 ptr2
   where
-    map_ :: Int -> Ptr Word8 -> Ptr Word8 -> IO ()
-    map_ !n !p1 !p2
-       | n >= len = return ()
-       | otherwise = do
-            x <- peekByteOff p1 n
-            pokeByteOff p2 n (f x)
-            map_ (n+1) p1 p2
-{-# INLINE map #-}
+    m p1 p2 = map_ 0
+      where
+      map_ :: Int -> IO ()
+      map_ !n
+         | n >= len = return ()
+         | otherwise = do
+              x <- peekByteOff p1 n
+              pokeByteOff p2 n (f x)
+              map_ (n+1)
```

Comment:
Looks like a problem with let-floating in the unsafe block. Nice catch.

Reply:
@Boarders, do you mind trying this version? In the wild I've found the Offset operations to be slower than just incrementing the pointer, since you do a little bit of redundant work constantly tracking the index and then incrementing the pointer to the offset by that amount per round (from start every time!). I'd like to catalogue where it stands against your implementation: 

```haskell
map :: (Word8 -> Word8) -> ByteString -> ByteString
map f (BS fp len) = unsafeDupablePerformIO $ unsafeWithForeignPtr fp $ \srcPtr ->
    create len $ \dstPtr -> m srcPtr dstPtr (srcPtr `plusPtr` len)
  where
    m !src !dst !end = map_ src dst
      where
      map_ :: Ptr Word8 -> Ptr Word8 -> IO ()
      map_ !p1 !p2
        | p1 == end = return ()
        | otherwise = do
          x <- peek p1
          poke p2 (f x)
          map_ (p1 `plusPtr` 1) (p2 `plusPtr` 1)
{-# INLINE map #-}
```

Reply:
Also i notice you're doing `unsafeWithForeignPtr` here with `create`. Is it worth making a change to `unsafeCreate` to use `unsafeWithForeignPtr` so we can just call that instead of `unsafeDupablePerformIO $ unsafeWithForeignPtr fp $ \srcPtr ->
    create...`?

In practice, this is the same for newer GHC's iirc

Reply:
@emilypi: Here are the benchmarks I get for that version:
with INLINE (so map is not inlined):
```
traversals/map (+1)                      mean 343.4 μs  ( +- 5.397 μs  )
traversals/map (+1)                      mean 85.40 ns  ( +- 1.411 ns  )
```

without inline:
```
traversals/map (+1)                      mean 107.6 μs  ( +- 2.118 μs  )
traversals/map (+1)                      mean 31.81 ns  ( +- 407.4 ps  )
```

So this does worse both inlined and not inlined. That is strange to me since I have the same intuition as you that using an offset is worse than directly incrementing a pointer. But perhaps because the offset is shared it does better.

Here is the core for this version:
```haskell
map :: (Word8 -> Word8) -> ByteString -> ByteString
map
  = \ (w_sg4O :: Word8 -> Word8) (w1_sg4P :: ByteString) ->
      case w1_sg4P of { BS ww1_sg4S ww2_sg4T ww3_sg4U ->
      $wmap w_sg4O ww1_sg4S ww2_sg4T ww3_sg4U
      }

$wmap
  :: (Word8 -> Word8)
     -> Addr# -> ForeignPtrContents -> Int# -> ByteString
$wmap
  = \ (w_sg4O :: Word8 -> Word8)
      (ww_sg4S :: Addr#)
      (ww1_sg4T :: ForeignPtrContents)
      (ww2_sg4U :: Int#) ->
      case runRW#
             (\ (s_i7pe :: State# RealWorld) ->
                case <# ww2_sg4U 0# of {
                  __DEFAULT ->
                    case newPinnedByteArray# ww2_sg4U s_i7pe of
                    { (# ipv_i6vK, ipv1_i6vL #) ->
                    let {
                      ipv2_s9JH :: Addr#
                      ipv2_s9JH = byteArrayContents# (ipv1_i6vL `cast` <Co:5>) } in
                    let {
                      end_s9JF :: Addr#
                      end_s9JF = plusAddr# ww_sg4S ww2_sg4U } in
                    join {
                      $w$j_sg4B :: State# RealWorld -> (# State# RealWorld, ByteString #)
                      $w$j_sg4B (w1_sg4z :: State# RealWorld)
                        = let {
                            ipv3_s9JG :: ForeignPtrContents
                            ipv3_s9JG = PlainPtr ipv1_i6vL } in
                          case touch# ipv3_s9JG w1_sg4z of s'_i6w0 { __DEFAULT ->
                          case touch# ww1_sg4T s'_i6w0 of s'1_i7pr { __DEFAULT ->
                          (# s'1_i7pr, BS ipv2_s9JH ipv3_s9JG ww2_sg4U #)
                          }
                          } } in
                    joinrec {
                      $wmap__sg4N
                        :: Addr#
                           -> Addr# -> State# RealWorld -> (# State# RealWorld, ByteString #)
                      $wmap__sg4N (ww3_sg4H :: Addr#)
                                  (ww4_sg4L :: Addr#)
                                  (w1_sg4E :: State# RealWorld)
                        = case eqAddr# ww3_sg4H end_s9JF of {
                            __DEFAULT ->
                              case readWord8OffAddr# ww3_sg4H 0# w1_sg4E of
                              { (# ipv4_i7tV, ipv5_i7tW #) ->
                              case w_sg4O (W8# ipv5_i7tW) of { W8# x_i7uf ->
                              case writeWord8OffAddr# ww4_sg4L 0# x_i7uf ipv4_i7tV of s2_i7uh
                              { __DEFAULT ->
                              jump $wmap__sg4N
                                (plusAddr# ww3_sg4H 1#) (plusAddr# ww4_sg4L 1#) s2_i7uh
                              }
                              }
                              };
                            1# -> jump $w$j_sg4B w1_sg4E
                          }; } in
                    jump $wmap__sg4N ww_sg4S ipv2_s9JH ipv_i6vK
                    };
                  1# -> case mallocPlainForeignPtrBytes2 of wild_00 { }
                })
      of
      { (# ipv_i6Ao, ipv1_i6Ap #) ->
      ipv1_i6Ap
      }
```

Reply:
Ah it seems to me that the inner loop is not inlining well with `m`. It may require a pragma.

Reply:
Pointer arithmetic is extremely fast on CPU, much faster than storing an additional argument on the stack (saying nothing of heap).

Reply:
Here's the core for the reworked version (maximal sharing). If you squint, you can see two places where it goes wrong - unnecessary touching with each increment, and in `eqAddr#`, which I imagine does not optimize as well as it's not `Bool#` valued. 

```haskell
 letrec {
                                 map__a60q [Occ=LoopBreaker] :: Ptr Word8 -> Ptr Word8 -> IO ()
                                 [LclId, Arity=2, Unf=OtherCon []]
                                 map__a60q
                                   = \ (src_a46A [Occ=Once!] :: Ptr Word8)
                                       (dst_a46B [Occ=Once!] :: Ptr Word8) ->
                                       case src_a46A of { GHC.Ptr.Ptr ipv4_s7uU ->
                                       case dst_a46B of { GHC.Ptr.Ptr ipv5_s7uX ->
                                       case GHC.Prim.eqAddr# ipv4_s7uU end_s8lZ of {
                                         __DEFAULT ->
                                           (\ (s1_i7pQ [Occ=Once]
                                                 :: GHC.Prim.State# GHC.Prim.RealWorld) ->
                                              case GHC.Prim.readWord8OffAddr#
                                                     @ GHC.Prim.RealWorld ipv4_s7uU 0# s1_i7pQ
                                              of
                                              { (# ipv6_i7tT [Occ=Once], ipv7_i7tU [Occ=Once] #) ->
                                              case f_a46t (GHC.Word.W8# ipv7_i7tU) of
                                              { W8# x_i7ud [Occ=Once] ->
                                              case GHC.Prim.writeWord8OffAddr#
                                                     @ GHC.Prim.RealWorld
                                                     ipv5_s7uX
                                                     0#
                                                     x_i7ud
                                                     ipv6_i7tT
                                              of s2_i7uf [Occ=Once]
                                              { __DEFAULT ->
                                              ((map__a60q
                                                  (GHC.Ptr.Ptr
                                                     @ Word8 (GHC.Prim.plusAddr# ipv4_s7uU 1#))
                                                  (GHC.Ptr.Ptr
                                                     @ Word8 (GHC.Prim.plusAddr# ipv5_s7uX 1#)))
                                               `cast` (GHC.Types.N:IO[0] <()>_R
                                                       :: IO ()
                                                          ~R# (GHC.Prim.State# GHC.Prim.RealWorld
                                                               -> (# GHC.Prim.State#
                                                                       GHC.Prim.RealWorld,
                                                                     () #))))
                                                s2_i7uf
                                              }
                                              }
                                              })
                                           `cast` (Sym (GHC.Types.N:IO[0] <()>_R)
                                                   :: (GHC.Prim.State# GHC.Prim.RealWorld
                                                       -> (# GHC.Prim.State# GHC.Prim.RealWorld,
                                                             () #))
                                                      ~R# IO ());
                                         1# ->
                                           (\ (s1_i7r0 [Occ=Once]
                                                 :: GHC.Prim.State# GHC.Prim.RealWorld) ->
                                              (# s1_i7r0, GHC.Tuple.() #))
                                           `cast` (Sym (GHC.Types.N:IO[0] <()>_R)
                                                   :: (GHC.Prim.State# GHC.Prim.RealWorld
                                                       -> (# GHC.Prim.State# GHC.Prim.RealWorld,
                                                             () #))
                                                      ~R# IO ())
                                       }
                                       }
                                       };
```

Reply:
I'm happy with your version as a result :)